pkg/roll/farm.go

```diff
@@ -0,0 +1,203 @@
+package roll
+
+import (
+	"github.com/square/p2/Godeps/_workspace/src/github.com/Sirupsen/logrus"
+
+	"github.com/square/p2/pkg/health/checker"
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/rcstore"
+	"github.com/square/p2/pkg/kp/rollstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/rc"
+	"github.com/square/p2/pkg/rc/fields"
+	roll_fields "github.com/square/p2/pkg/roll/fields"
+)
+
+type Factory interface {
+	New(roll_fields.Update, logging.Logger, kp.Lock) Update
+}
+
+type UpdateFactory struct {
+	KPStore       kp.Store
+	RCStore       rcstore.Store
+	HealthChecker checker.ConsulHealthChecker
+	Labeler       labels.Applicator
+	Scheduler     rc.Scheduler
+}
+
+func (f UpdateFactory) New(u roll_fields.Update, l logging.Logger, lock kp.Lock) Update {
+	return NewUpdate(u, f.KPStore, f.RCStore, f.HealthChecker, f.Labeler, f.Scheduler, l, lock)
+}
+
+// The Farm is responsible for spawning and reaping rolling updates as they are
+// added to and deleted from Consul. Multiple farms can exist simultaneously,
+// but each one must hold a different Consul session. This ensures that the
+// farms do not instantiate the same rolling update multiple times.
+type Farm struct {
+	factory  Factory
+	kps      kp.Store
+	rls      rollstore.Store
+	sessions <-chan string
+
+	children map[fields.ID]childRU
+	lock     *kp.Lock
+
+	logger logging.Logger
+}
+
+type childRU struct {
+	ru   Update
+	quit chan<- struct{}
+}
+
+func NewFarm(
+	factory Factory,
+	kps kp.Store,
+	rls rollstore.Store,
+	sessions <-chan string,
+	logger logging.Logger,
+) *Farm {
+	return &Farm{
+		factory:  factory,
+		kps:      kps,
+		rls:      rls,
+		sessions: sessions,
+		logger:   logger,
+		children: make(map[fields.ID]childRU),
+	}
+}
+
+// Start is a blocking function that monitors Consul for updates. The Farm will
+// attempt to claim updates as they appear and, if successful, will start
+// goroutines for those updatesto do their job. Closing the quit channel will
+// cause this function to return, releasing all locks it holds.
+//
+// Start is not safe for concurrent execution. Do not execute multiple
+// concurrent instances of Start.
```

Comment:
I don't think enforcing the single-start is important, I just wanted to document it because there's shared state in the Farm object itself. Personally, I think the entire function is already approaching its maximum allowable complexity, and there's not much value to be gained by checking the single-start-ness.